src/bjointsp/read_write/writer.py

```diff
@@ -101,6 +103,12 @@ def save_heuristic_variables(result, changed_instances, instances, edges, nodes,
                     consumed_dr[(path[i], path[i+1])] += e.flow_dr() / len(e.paths)
                     link = {"arc": str(e.arc), "edge_src": e.source.location, "edge_dst": e.dest.location, "link_src": path[i], "link_dst": path[i+1]}
                     result["placement"]["links"].append(link)
+        for path in e.paths:
+            # record vnf delays
+            vnf_delay = {"dest_vnf": e.arc.dest.name, "dest_vnf_delay": e.arc.dest.vnf_delay }
+            result["metrics"]["vnf_delays"].append(vnf_delay)
+            result["metrics"]["total_vnf_delay"] += e.arc.dest.vnf_delay
+
```

Comment:
Nice, the `total_path_delay` and `total_vnf_delay` look good. To minor suggestions:

* In the `vnf_delays` section, simply call the fields `vnf` and `vnf_delay` instead of `dest_vnf` and `dest_vnf_delay`. I think it's a bit clearer when looking at the result file.
* Could you add a third field `total_delay = total_path_delay + total_vnf_delay` for convenience?